File: crossword_generator.php
<?php
// crossword_generator.php — PHP версия

class MiniCrossword {
    private $size;
    private $grid;
    private $solution;
    private $words;
    private $placedWords;
    private $wordPositions;

    public function __construct($size = 5, $seed = null) {
        $this->size = $size;
        $this->grid = array_fill(0, $size, array_fill(0, $size, ' '));
        $this->solution = array_fill(0, $size, array_fill(0, $size, ' '));
        $this->words = ['АРБУЗ', 'БУКЕТ', 'ГАМА', 'ДОМ', 'ЕЛЬ', 'ЁЖИК', 'ЖУК', 'ЗОНТ', 'ИГЛА'];
        $this->placedWords = [];
        $this->wordPositions = [];
        if ($seed !== null) mt_srand($seed);
    }

    public function generate() {
        $sortedWords = $this->words;
        usort($sortedWords, function($a, $b) {
            return strlen($b) - strlen($a);
        });

        foreach ($sortedWords as $word) {
            if ($this->placeWord($word)) {
                $this->placedWords[] = $word;
            }
        }

        for ($i = 0; $i < $this->size; $i++) {
            for ($j = 0; $j < $this->size; $j++) {
                $this->solution[$i][$j] = $this->grid[$i][$j];
            }
        }

        $letters = str_split('АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ');
        for ($i = 0; $i < $this->size; $i++) {
            for ($j = 0; $j < $this->size; $j++) {
                if ($this->grid[$i][$j] == ' ') {
                    $this->grid[$i][$j] = $letters[array_rand($letters)];
                }
            }
        }
    }

    private function placeWord($word) {
        $dirs = [[0, 1], [1, 0]];
        for ($attempt = 0; $attempt < 50; $attempt++) {
            $row = rand(0, $this->size - 1);
            $col = rand(0, $this->size - 1);
            $dir = $dirs[array_rand($dirs)];
            if ($this->canPlace($word, $row, $col, $dir)) {
                $this->placeWordInGrid($word, $row, $col, $dir);
                $this->wordPositions[$word] = [$row, $col, $dir];
                return true;
            }
        }
        return false;
    }

    private function canPlace($word, $row, $col, $dir) {
        list($dr, $dc) = $dir;
        if ($row + $dr * (strlen($word) - 1) >= $this->size || $col + $dc * (strlen($word) - 1) >= $this->size) {
            return false;
        }
        $chars = str_split($word);
        foreach ($chars as $i => $ch) {
            $r = $row + $dr * $i;
            $c = $col + $dc * $i;
            if ($this->grid[$r][$c] != ' ' && $this->grid[$r][$c] != $ch) {
                return false;
            }
        }
        return true;
    }

    private function placeWordInGrid($word, $row, $col, $dir) {
        list($dr, $dc) = $dir;
        $chars = str_split($word);
        foreach ($chars as $i => $ch) {
            $r = $row + $dr * $i;
            $c = $col + $dc * $i;
            $this->grid[$r][$c] = $ch;
        }
    }

    private function getNumber($row, $col) {
        $num = 1;
        foreach ($this->wordPositions as $pos) {
            list($r, $c, $dir) = $pos;
            if ($r == $row && $c == $col) {
                return $num;
            }
            $num++;
        }
        return 0;
    }

    public function printCrossword($showSolution = false) {
        $grid = $showSolution ? $this->solution : $this->grid;
        echo "┌─────┬─────┬─────┬─────┬─────┐\n";
        for ($i = 0; $i < $this->size; $i++) {
            echo "│";
            for ($j = 0; $j < $this->size; $j++) {
                if ($grid[$i][$j] == ' ') {
                    echo "   ";
                } else {
                    $num = $this->getNumber($i, $j);
                    if ($num > 0 && !$showSolution) {
                        echo str_pad($num, 2, " ", STR_PAD_LEFT) . " ";
                    } else {
                        echo " {$grid[$i][$j]} ";
                    }
                }
                if ($j < $this->size - 1) echo "│";
            }
            echo "│\n";
            if ($i < $this->size - 1) {
                echo "├─────┼─────┼─────┼─────┼─────┤\n";
            }
        }
        echo "└─────┴─────┴─────┴─────┴─────┘\n";
    }

    public function printClues() {
        echo "\nСлова по горизонтали:\n";
        foreach ($this->placedWords as $i => $word) {
            if ($this->wordPositions[$word][2][0] == 0 && $this->wordPositions[$word][2][1] == 1) {
                echo ($i + 1) . ". $word\n";
            }
        }
        echo "\nСлова по вертикали:\n";
        foreach ($this->placedWords as $i => $word) {
            if ($this->wordPositions[$word][2][0] == 1 && $this->wordPositions[$word][2][1] == 0) {
                echo ($i + 1) . ". $word\n";
            }
        }
    }

    public function saveJSON($filename = 'crossword.json') {
        $data = [
            'size' => $this->size,
            'grid' => $this->grid,
            'solution' => $this->solution,
            'words' => $this->placedWords
        ];
        file_put_contents($filename, json_encode($data, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
        echo "💾 Сохранено: $filename\n";
    }

    public function saveHTML($filename = 'crossword.html') {
        $html = '<!DOCTYPE html>
<html>
<head><meta charset="UTF-8"><title>Кроссворд</title>
<style>
body { font-family: monospace; background: #1a1a2e; color: #fff; padding: 20px; }
table { border-collapse: collapse; margin: 20px auto; }
td { width: 40px; height: 40px; text-align: center; font-size: 18px; border: 2px solid #444; background: #2a2a4e; }
td.filled { background: #3a3a6e; }
td.empty { background: #1a1a2e; }
.number { font-size: 10px; color: #888; position: relative; top: -15px; left: -15px; }
</style>
</head>
<body>
<h1 style="text-align:center;">🧩 Кроссворд</h1>
<table>
';
        for ($i = 0; $i < $this->size; $i++) {
            $html .= "<tr>";
            for ($j = 0; $j < $this->size; $j++) {
                $cell = $this->grid[$i][$j];
                $cls = $cell == ' ' ? 'empty' : 'filled';
                $num = $this->getNumber($i, $j);
                $display = $cell == ' ' ? '&nbsp;' : $cell;
                $numHtml = $num > 0 ? "<span class=\"number\">$num</span>" : '';
                $html .= "<td class=\"$cls\">$numHtml$display</td>";
            }
            $html .= "</tr>\n";
        }
        $html .= '</table>
<p style="text-align:center;">💡 Подсказки доступны в консоли</p>
</body>
</html>';
        file_put_contents($filename, $html);
        echo "💾 Сохранено: $filename\n";
    }
}

function main() {
    echo "🧩 Mini Crossword Generator (PHP)\n";
    echo "Размер кроссворда (5 или 7): ";
    $sizeInput = trim(fgets(STDIN));
    $size = empty($sizeInput) ? 5 : (int)$sizeInput;
    $size = in_array($size, [5, 7]) ? $size : 5;

    echo "Seed (число, опционально): ";
    $seedInput = trim(fgets(STDIN));
    $seed = empty($seedInput) ? null : (int)$seedInput;

    $gen = new MiniCrossword($size, $seed);
    $gen->generate();

    echo "\nКроссворд:\n";
    $gen->printCrossword();

    echo "\nПоказать решение? (y/n): ";
    $show = trim(fgets(STDIN));
    if (strtolower($show) == 'y') {
        echo "\nРешение:\n";
        $gen->printCrossword(true);
    }

    $gen->printClues();
    $gen->saveJSON();
    $gen->saveHTML();
}

main();
?>

File: crossword_generator.rs
// crossword_generator.rs — Rust версия

use rand::seq::SliceRandom;
use rand::thread_rng;
use std::collections::HashMap;
use std::fs;
use std::io::{self, Write};

struct MiniCrossword {
    size: usize,
    grid: Vec<Vec<char>>,
    solution: Vec<Vec<char>>,
    words: Vec<String>,
    placed_words: Vec<String>,
    word_positions: HashMap<String, (usize, usize, (isize, isize))>,
    rng: rand::rngs::ThreadRng,
}

impl MiniCrossword {
    fn new(size: usize) -> Self {
        let grid = vec![vec![' '; size]; size];
        let solution = vec![vec![' '; size]; size];
        let words = vec![
            "АРБУЗ".to_string(), "БУКЕТ".to_string(), "ГАМА".to_string(),
            "ДОМ".to_string(), "ЕЛЬ".to_string(), "ЁЖИК".to_string(),
            "ЖУК".to_string(), "ЗОНТ".to_string(), "ИГЛА".to_string()
        ];
        MiniCrossword {
            size,
            grid,
            solution,
            words,
            placed_words: Vec::new(),
            word_positions: HashMap::new(),
            rng: thread_rng(),
        }
    }

    fn generate(&mut self) {
        let mut sorted_words = self.words.clone();
        sorted_words.sort_by(|a, b| b.len().cmp(&a.len()));

        for word in sorted_words {
            if self.place_word(&word) {
                self.placed_words.push(word);
            }
        }

        for i in 0..self.size {
            for j in 0..self.size {
                self.solution[i][j] = self.grid[i][j];
            }
        }

        let letters = "АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ";
        let letters_vec: Vec<char> = letters.chars().collect();
        for i in 0..self.size {
            for j in 0..self.size {
                if self.grid[i][j] == ' ' {
                    self.grid[i][j] = *letters_vec.choose(&mut self.rng).unwrap();
                }
            }
        }
    }

    fn place_word(&mut self, word: &str) -> bool {
        let dirs = [(0, 1), (1, 0)];
        let chars: Vec<char> = word.chars().collect();
        for _ in 0..50 {
            let row = self.rng.gen_range(0..self.size);
            let col = self.rng.gen_range(0..self.size);
            let dir = *dirs.choose(&mut self.rng).unwrap();
            if self.can_place(&chars, row, col, dir) {
                self.place_word_in_grid(&chars, row, col, dir);
                self.word_positions.insert(word.to_string(), (row, col, dir));
                return true;
            }
        }
        false
    }

    fn can_place(&self, chars: &[char], row: usize, col: usize, dir: (isize, isize)) -> bool {
        let (dr, dc) = dir;
        if (row as isize + dr * (chars.len() as isize - 1)) >= self.size as isize ||
           (col as isize + dc * (chars.len() as isize - 1)) >= self.size as isize {
            return false;
        }
        for (i, &ch) in chars.iter().enumerate() {
            let r = (row as isize + dr * i as isize) as usize;
            let c = (col as isize + dc * i as isize) as usize;
            if self.grid[r][c] != ' ' && self.grid[r][c] != ch {
                return false;
            }
        }
        true
    }

    fn place_word_in_grid(&mut self, chars: &[char], row: usize, col: usize, dir: (isize, isize)) {
        let (dr, dc) = dir;
        for (i, &ch) in chars.iter().enumerate() {
            let r = (row as isize + dr * i as isize) as usize;
            let c = (col as isize + dc * i as isize) as usize;
            self.grid[r][c] = ch;
        }
    }

    fn get_number(&self, row: usize, col: usize) -> usize {
        let mut num = 1;
        for (_, &(r, c, _)) in self.word_positions.iter() {
            if r == row && c == col {
                return num;
            }
            num += 1;
        }
        0
    }

    fn print_crossword(&self, show_solution: bool) {
        let grid = if show_solution { &self.solution } else { &self.grid };
        println!("┌─────┬─────┬─────┬─────┬─────┐");
        for i in 0..self.size {
            print!("│");
            for j in 0..self.size {
                if grid[i][j] == ' ' {
                    print!("   ");
                } else {
                    let num = self.get_number(i, j);
                    if num > 0 && !show_solution {
                        print!("{:2} ", num);
                    } else {
                        print!(" {} ", grid[i][j]);
                    }
                }
                if j < self.size - 1 {
                    print!("│");
                }
            }
            println!("│");
            if i < self.size - 1 {
                println!("├─────┼─────┼─────┼─────┼─────┤");
            }
        }
        println!("└─────┴─────┴─────┴─────┴─────┘");
    }

    fn print_clues(&self) {
        println!("\nСлова по горизонтали:");
        for (i, word) in self.placed_words.iter().enumerate() {
            if let Some(&(_, _, dir)) = self.word_positions.get(word) {
                if dir == (0, 1) {
                    println!("{}. {}", i + 1, word);
                }
            }
        }
        println!("\nСлова по вертикали:");
        for (i, word) in self.placed_words.iter().enumerate() {
            if let Some(&(_, _, dir)) = self.word_positions.get(word) {
                if dir == (1, 0) {
                    println!("{}. {}", i + 1, word);
                }
            }
        }
    }

    fn save_json(&self, filename: &str) {
        let data = serde_json::json!({
            "size": self.size,
            "grid": self.grid,
            "solution": self.solution,
            "words": self.placed_words,
        });
        let json = serde_json::to_string_pretty(&data).unwrap();
        fs::write(filename, json).unwrap();
        println!("💾 Сохранено: {}", filename);
    }

    fn save_html(&self, filename: &str) {
        let mut html = String::from(r#"<!DOCTYPE html>
<html>
<head><meta charset="UTF-8"><title>Кроссворд</title>
<style>
body { font-family: monospace; background: #1a1a2e; color: #fff; padding: 20px; }
table { border-collapse: collapse; margin: 20px auto; }
td { width: 40px; height: 40px; text-align: center; font-size: 18px; border: 2px solid #444; background: #2a2a4e; }
td.filled { background: #3a3a6e; }
td.empty { background: #1a1a2e; }
.number { font-size: 10px; color: #888; position: relative; top: -15px; left: -15px; }
</style>
</head>
<body>
<h1 style="text-align:center;">🧩 Кроссворд</h1>
<table>
"#);
        for i in 0..self.size {
            html.push_str("<tr>");
            for j in 0..self.size {
                let cell = self.grid[i][j];
                let cls = if cell == ' ' { "empty" } else { "filled" };
                let num = self.get_number(i, j);
                let display = if cell == ' ' { "&nbsp;" } else { cell.to_string() };
                let num_html = if num > 0 { format!("<span class=\"number\">{}</span>", num) } else { "".to_string() };
                html.push_str(&format!("<td class=\"{}\">{}{}</td>", cls, num_html, display));
            }
            html.push_str("</tr>\n");
        }
        html.push_str(r#"</table>
<p style="text-align:center;">💡 Подсказки доступны в консоли</p>
</body>
</html>"#);
        fs::write(filename, html).unwrap();
        println!("💾 Сохранено: {}", filename);
    }
}

fn main() {
    println!("🧩 Mini Crossword Generator (Rust)");
    print!("Размер кроссворда (5 или 7): ");
    io::stdout().flush().unwrap();
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let size = input.trim().parse::<usize>().unwrap_or(5);
    let size = if size == 5 || size == 7 { size } else { 5 };

    let mut gen = MiniCrossword::new(size);
    gen.generate();

    println!("\nКроссворд:");
    gen.print_crossword(false);

    print!("\nПоказать решение? (y/n): ");
    io::stdout().flush().unwrap();
    input.clear();
    io::stdin().read_line(&mut input).unwrap();
    if input.trim().to_lowercase() == "y" {
        println!("\nРешение:");
        gen.print_crossword(true);
    }

    gen.print_clues();
    gen.save_json("crossword.json");
    gen.save_html("crossword.html");
}
